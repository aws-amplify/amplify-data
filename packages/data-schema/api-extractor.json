{
  "extends": "../../api-extractor.base.json",
  "mainEntryPointFilePath": "<projectFolder>/dist/esm/src/index.d.ts"
}
