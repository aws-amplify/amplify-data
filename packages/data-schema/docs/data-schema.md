<!-- Do not edit this file. It is automatically generated by API Documenter. -->

[Home](./index.md) &gt; [@aws-amplify/data-schema](./data-schema.md)

## data-schema package

## Enumerations

<table><thead><tr><th>

Enumeration


</th><th>

Description


</th></tr></thead>
<tbody><tr><td>

[ModelRelationshipTypes](./data-schema.modelrelationshiptypes.md)


</td><td>

Model relationship types


</td></tr>
</tbody></table>

## Interfaces

<table><thead><tr><th>

Interface


</th><th>

Description


</th></tr></thead>
<tbody><tr><td>

[EnumType](./data-schema.enumtype.md)


</td><td>

Enum type definition content


</td></tr>
</tbody></table>

## Namespaces

<table><thead><tr><th>

Namespace


</th><th>

Description


</th></tr></thead>
<tbody><tr><td>

[a](./data-schema.a.md)


</td><td>


</td></tr>
</tbody></table>

## Type Aliases

<table><thead><tr><th>

Type Alias


</th><th>

Description


</th></tr></thead>
<tbody><tr><td>

[Authorization](./data-schema.authorization.md)


</td><td>

Container for authorization schema definition content.


</td></tr>
<tr><td>

[ClientSchema](./data-schema.clientschema.md)


</td><td>


</td></tr>
<tr><td>

[CombinedModelSchema](./data-schema.combinedmodelschema.md)


</td><td>

CombinedModel schema definition interface


</td></tr>
<tr><td>

[CustomOperation](./data-schema.customoperation.md)


</td><td>

Custom operation definition interface


</td></tr>
<tr><td>

[CustomType](./data-schema.customtype.md)


</td><td>

Custom type container


</td></tr>
<tr><td>

[Json](./data-schema.json.md)


</td><td>

A precise, recursive Json type blows the type calculation stack without installing explicit `Json extends T ? short-circuit : ...` type checks all over the place. We may take that on later. But, this is a good-enough approximation for now.


</td></tr>
<tr><td>

[ModelDefaultIdentifier](./data-schema.modeldefaultidentifier.md)


</td><td>

Model default identifier


</td></tr>
<tr><td>

[ModelField](./data-schema.modelfield.md)


</td><td>

Public API for the chainable builder methods exposed by Model Field. The type is narrowing e.g., after calling .array() it will be omitted from intellisense suggestions


</td></tr>
<tr><td>

[ModelRelationshipField](./data-schema.modelrelationshipfield.md)


</td><td>

Model relationship field definition interface


</td></tr>
<tr><td>

[ModelRelationshipTypeArgFactory](./data-schema.modelrelationshiptypeargfactory.md)


</td><td>

Model relationship type definition content


</td></tr>
<tr><td>

[ModelSchema](./data-schema.modelschema.md)


</td><td>

Model schema definition interface


</td></tr>
<tr><td>

[ModelType](./data-schema.modeltype.md)


</td><td>

Model type definition interface


</td></tr>
<tr><td>

[Nullable](./data-schema.nullable.md)


</td><td>

Field type arg mutators


</td></tr>
<tr><td>

[RDSModelSchema](./data-schema.rdsmodelschema.md)


</td><td>

RDSModel schema definition interface


</td></tr>
<tr><td>

[RefType](./data-schema.reftype.md)


</td><td>

Reference type definition interface


</td></tr>
<tr><td>

[ResourceAuthorization](./data-schema.resourceauthorization.md)


</td><td>

Container for representing the authorization of function resources


</td></tr>
</tbody></table>
