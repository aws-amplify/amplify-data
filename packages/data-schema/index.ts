import * as a from './src';
import { ClientSchema } from './src/ClientSchema';

export { a };

export type { ClientSchema };

export * from './src/runtime/client';
export {
  addSchemaToClient,
  addSchemaToClientWithInstance,
} from './src/runtime';
