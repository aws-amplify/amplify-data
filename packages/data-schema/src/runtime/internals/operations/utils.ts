// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0
// import { GraphQLFormattedError } from '@aws-amplify/data-schema-types';

/**
 * Handle errors for list return types (list and index query operations)
 */
export function handleListGraphQlError(error: any) {
  if (error?.errors) {
    // graphql errors pass through
    return {
      ...error,
      data: [],
    } as any;
  } else {
    // non-graphql errors are re-thrown
    throw error;
  }
}

/**
 * Handle errors for singular return types (create, get, update, delete operations)
 */
export function handleSingularGraphQlError(error: any) {
  if (error.errors) {
    // graphql errors pass through
    return {
      ...error,
      data: null,
    } as any;
  } else {
    // non-graphql errors are re-thrown
    throw error;
  }
}

/**
 *
 * @param resolver
 * @returns
 */
export function selfAwareAsync<T>(
  resolver: (promise: Promise<T>) => Promise<T>,
): Promise<T> {
  let resolve: (value: T) => void;
  let reject: (reason?: any) => void;

  const resultPromise = new Promise<T>((res, rej) => {
    resolve = res;
    reject = rej;
  });

  resolver(resultPromise)
    .then((result) => {
      resolve(result);
    })
    .catch((error) => {
      reject(error);
    });

  return resultPromise;
}
