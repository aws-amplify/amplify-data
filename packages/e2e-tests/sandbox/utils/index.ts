import { deploySandbox, deploySandboxWatch } from './deploy';
import { teardownSandbox } from './teardown';
import { updateSchema } from './schemaUtils';

/**
 * Sandbox operations can sometimes be time consuming, so we increase the Jest
 * setup / teardown hook timeouts to 10 minutes.
 */
export const sandboxTimeout: number = 600000;

export const outputsFileName = 'amplify_outputs.json';

export { deploySandbox, deploySandboxWatch };
export { teardownSandbox };
export { updateSchema };
