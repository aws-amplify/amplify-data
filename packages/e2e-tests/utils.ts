import { Amplify } from 'aws-amplify';
import { generateClient } from 'aws-amplify/data';
import type { Schema } from './amplify/data/resource';
import outputs from './amplify_outputs.json';

export const statuses = {
  success: {
    icon: '✅',
    label: 'Success',
  },
  fail: {
    icon: '❌',
    label: 'Fail',
  },
} as const;

const _testCase = (success: boolean): any =>
  success ? statuses.success : statuses.fail;

// TODO: use imported type from `aws-amplify/data` once it's fixed
type Client = ReturnType<typeof configureAmplifyAndGenerateClient>;

export type TestCase = {
  label: string;
  action: (client: Client) => Promise<boolean>;
};

/**
 * Given an array of test cases, runs each test case
 * and asserts that the result is successful
 * @param testCases
 * @param client
 */
export async function runTestCases(testCases: TestCase[]) {
  const client = (global as any).client as Client;

  for (const testCase of testCases) {
    test(`${testCase.label}`, async () => {
      let result: any;

      try {
        const response = await testCase.action(client);
        result = _testCase(response);
      } catch (error) {
        result = _testCase(false);
      }

      console.log(`${result.icon} ${testCase.label}`);
      expect(result.label).toBe(statuses.success.label);
    });
  }
}

/**
 * Configures Amplify and returns API client
 * @returns API client
 */
export function configureAmplifyAndGenerateClient() {
  console.log('configuring Amplify and generating client..');
  Amplify.configure(outputs);

  const client = generateClient<Schema>();

  return client;
}
