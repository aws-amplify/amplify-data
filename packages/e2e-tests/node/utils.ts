import { Amplify } from 'aws-amplify';
import {
  CONNECTION_STATE_CHANGE,
  ConnectionState,
  generateClient,
} from 'aws-amplify/data';
import { Hub, ConsoleLogger } from 'aws-amplify/utils';
import type { Schema } from './amplify/data/resource';
import outputs from './amplify_outputs.json';
// TODO: the dep exists, investigate why it's not being resolved:
import { WebSocket } from 'ws';

// TODO: use imported type from `aws-amplify/data` once it's fixed
export type Client = ReturnType<typeof generateClient<Schema>>;

type ConfigureAmplifyAndGenerateClientParams = {
  disableDebugLogging?: boolean;
  amplifyOptions?: any;
  apiClientOptions?: any;
};

/**
 * Configures Amplify and returns API client
 * @param disableDebugLogging - disables debug logging
 * @param amplifyOptions - options to pass to Amplify.configure
 * @param apiClientOptions - options to pass to generateClient
 * @returns API client
 */
export const configureAmplifyAndGenerateClient = ({
  disableDebugLogging = false,
  amplifyOptions = {},
  apiClientOptions = {},
}: ConfigureAmplifyAndGenerateClientParams): Client => {
  console.log('configuring Amplify and generating client..');
  Amplify.configure(outputs, amplifyOptions);

  /**
   * Debug logs are enabled by default to assist with CI failure debugging.
   * Option to disable is included for local development, in the event that
   * the debug logs are too verbose.
   */
  if (!disableDebugLogging) {
    ConsoleLogger.LOG_LEVEL = 'DEBUG';

    Hub.listen('core', (data: any) => {
      if (data.payload.event === 'configure') {
        console.log('API configuration details:', data.payload.data.API);
      }
    });
  }

  return generateClient<Schema>(apiClientOptions);
};

/**
 * Util that takes model operation response and throws an error if errors
 * are present, or if the data is undefined.
 * @param response - model operation response
 * @param operation - operation name
 * @returns data from response
 */
export async function pause(ms: number) {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

type EstablishWebsocketParams = {
  disableConnectionStateLogging?: boolean;
};

/**
 * Adds the WebSocket API globally.
 * Subscriptions do not work in Node.js environment without the WebSocket API.
 */
export const establishWebsocket = ({
  disableConnectionStateLogging = false,
}: EstablishWebsocketParams): any => {
  (global as any).WebSocket = WebSocket;

  /**
   * Debug logs are enabled by default to assist with CI failure debugging.
   * Option to disable is included for local development, in the event that
   * the debug logs are too verbose.
   */
  if (!disableConnectionStateLogging) {
    // https://docs.amplify.aws/gen1/javascript/build-a-backend/graphqlapi/subscribe-data/#subscription-connection-status-updates
    Hub.listen('api', (data: any) => {
      const { payload } = data;
      if (payload.event === CONNECTION_STATE_CHANGE) {
        const connectionState = payload.data.connectionState as ConnectionState;
        console.log(
          'Connection state has changed. Connection state: ',
          connectionState,
        );
      }
    });
  }
};

export const handleErrorsAndData = <T>(
  response: { data?: T; errors?: any },
  operation: string,
): T => {
  if (response.errors) {
    console.log(`error on ${operation}:`, response.errors);
    throw new Error(JSON.stringify(response.errors));
  }

  if (!response.data) {
    throw new Error(`no response data for ${operation}`);
  }

  return response.data;
};
