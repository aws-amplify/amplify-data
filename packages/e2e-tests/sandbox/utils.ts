import { Amplify } from 'aws-amplify';
import {
  CONNECTION_STATE_CHANGE,
  ConnectionState,
  generateClient,
} from 'aws-amplify/data';
import { Hub, ConsoleLogger } from 'aws-amplify/utils';
import type { Schema } from './amplify/data/resource';
import outputs from './amplify_outputs.json';
import { WebSocket } from 'ws';

// TODO: use imported type from `aws-amplify/data` once it's fixed
export type Client = ReturnType<typeof generateClient<Schema>>;

type ConfigureAmplifyAndGenerateClientParams = {
  enableDebugLogging?: boolean;
  amplifyOptions?: any;
  apiClientOptions?: any;
};

/**
 * Configures Amplify and returns API client
 * @param enableDebugLogging - enables debug logging
 * @param amplifyOptions - options to pass to Amplify.configure
 * @param apiClientOptions - options to pass to generateClient
 * @returns API client
 */
export const configureAmplifyAndGenerateClient = ({
  enableDebugLogging = false,
  amplifyOptions = {},
  apiClientOptions = {},
}: ConfigureAmplifyAndGenerateClientParams): Client => {
  console.log('configuring Amplify and generating client..');
  Amplify.configure(outputs, amplifyOptions);

  if (enableDebugLogging) {
    ConsoleLogger.LOG_LEVEL = 'DEBUG';

    Hub.listen('core', (data: any) => {
      if (data.payload.event === 'configure') {
        console.log('API configuration details:', data.payload.data.API);
      }
    });
  }

  return generateClient<Schema>(apiClientOptions);
};

type AddWebsocketPolyfillParams = {
  enableConnectionStateLogging?: boolean;
};

/**
 * Adds the WebSocket API globally.
 * Subscriptions do not work in Node.js environment without the WebSocket API.
 */
export const addWebsocketPolyfill = ({
  enableConnectionStateLogging = false,
}: AddWebsocketPolyfillParams) => {
  (global as any).WebSocket = WebSocket;

  if (enableConnectionStateLogging) {
    Hub.listen('api', (data: any) => {
      const { payload } = data;
      if (payload.event === CONNECTION_STATE_CHANGE) {
        const connectionState = payload.data.connectionState as ConnectionState;
        console.log(
          'Connection state has changed. Connection state: ',
          connectionState,
        );
      }
    });
  }
};

/**
 * Util that takes model operation response and throws an error if errors
 * are present, or if the data is undefined.
 * @param response - model operation response
 * @param operation - operation name
 * @returns data from response
 */
export const expectDataReturnWithoutErrors = <T>(
  response: { data?: T; errors?: any },
  operation: string,
): T => {
  if (response.errors) {
    console.log(`error on ${operation}:`, response.errors);
    throw new Error(JSON.stringify(response.errors));
  }

  if (!response.data) {
    throw new Error(`no response data for ${operation}`);
  }

  return response.data;
};

/**
 * Function that listens for "Subscription ack" event, and once it receives it,
 * resolves.
 * Jest will timeout if the test does not complete within the default timeout of
 * 5 seconds, so no timeout handling is required.
 */
export const waitForSubscriptionAck = () =>
  new Promise<void>((resolve) => {
    const cancel = Hub.listen('api', async (data: any) => {
      const { payload } = data;
      // TODO: update once we add export:
      if (payload.event === 'Subscription ack') {
        console.log('Subscription ack received:', payload.data);
        // Stop listening for msgs once we receive the ack:
        cancel();
        // Resolve so test execution can continue:
        resolve();
      }
    });
  });
