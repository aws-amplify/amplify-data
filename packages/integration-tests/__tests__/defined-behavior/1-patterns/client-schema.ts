import { a, ClientSchema } from '@aws-amplify/data-schema';
import { Amplify } from 'aws-amplify';
import { buildAmplifyConfig, mockedGenerateClient } from '../../utils';

describe('ClientSchema', () => {
  afterEach(() => {
    jest.clearAllMocks();
  });

  test('overview sample schema', () => {
    // #region covers 1f807200986a702a
    const schema = a
      .schema({
        Customer: a
          .model({
            customerId: a.id().required(),
            // fields can be of various scalar types,
            // such as string, boolean, float, integers etc.
            name: a.string(),
            // fields can be of custom types
            location: a.customType({
              // fields can be required or optional
              lat: a.float().required(),
              long: a.float().required(),
            }),
            // fields can be enums
            engagementStage: a.enum(['PROSPECT', 'INTERESTED', 'PURCHASED']),
            collectionId: a.id(),
            collection: a.belongsTo('Collection', 'collectionId'),
            // Use custom identifiers. By default, it uses an `id: a.id()` field
          })
          .identifier(['customerId']),
        Collection: a
          .model({
            customers: a.hasMany('Customer', 'collectionId'), // setup relationships between types
            tags: a.string().array(), // fields can be arrays
            representativeId: a.id().required(),
            // customize secondary indexes to optimize your query performance
          })
          .secondaryIndexes((index) => [index('representativeId')]),
      })
      .authorization((allow) => [allow.publicApiKey()]);
    // #endregion

    // just ensuring that the example schema code transforms correctly.
    expect(schema.transform().schema).toMatchSnapshot();
  });

  describe(`['ModelName']`, () => {
    describe(`['type']`, () => {
      const schema = a
        .schema({
          Post: a.model({
            title: a.string().required(),
            body: a.string().required(),
          }),
        })
        .authorization((allow) => [allow.publicApiKey()]);
      type Schema = ClientSchema<typeof schema>;

      beforeEach(async () => {
        const config = await buildAmplifyConfig(schema);
        Amplify.configure(config);
      });

      test(`matches scalar results`, async () => {
        const { generateClient } = mockedGenerateClient([{ data: null }]);
        const client = generateClient<Schema>();

        // We can grab the `Post` type from the Schema like this:
        type Post = Schema['Post']['type'];

        // We can perform scalar operations against client models as usual:
        const result = await client.models.Post.get({ id: 'abc' });

        // And then use the `Post` type to receive result data. Notably, graphql
        // scalar result data is always nullable. So we non-null assert (!) here.
        const _post: Post = result.data!;
      });

      test(`matches list-like results`, async () => {
        const { generateClient } = mockedGenerateClient([
          { data: { listPosts: [] } },
        ]);
        const client = generateClient<Schema>();

        // We can grab the `Post` type from the Schema like this:
        type Post = Schema['Post']['type'];

        // We can perform list-like operations against client models as usual:
        const result = await client.models.Post.list();

        // And then use the `Post` type to receive result data.
        const _posts: Post[] = result.data;
      });
    });
  });

  describe(`['CustomOperationName']`, () => {
    const schema = a.schema({
      MyQueryReturnType: a.customType({
        output: a.string(),
      }),
      myQuery: a
        .query()
        .arguments({
          input: a.string(),
        })
        .returns(
          a.customType({
            output: a.string(),
          }),
        ),
    });
    type Schema = ClientSchema<typeof schema>;

    /**
     * Just used to create a client with the correct type signature that will
     * not throw an error upon invocation. Not for checking runtime behavior.
     */
    async function getMockClient() {
      const { spy, generateClient } = mockedGenerateClient([{ data: null }]);
      const config = await buildAmplifyConfig(schema);
      Amplify.configure(config);
      const client = generateClient<Schema>();
      return client;
    }

    test(`['returnType'] matches the operation return type`, async () => {
      const client = await getMockClient();

      // We can grab the args and return types from the schema
      type myQueryArgs = Schema['myQuery']['args'];
      type myQueryResult = Schema['myQuery']['returnType'];

      // We can create strongly typed arguments ahead of time:
      const args: myQueryArgs = {
        // ctrl + space in VS Code auto-suggests `input: ...`
        input: 'some string',
      };

      // Our custom operation accepts the pre-built, typed args
      const result = await client.queries.myQuery(args);

      // And we can then use the return type for variables, useStates, etc.
      const out: myQueryResult = result.data;
    });

    test(`['functionHandler'] can be used add typing to lambda handlers`, async () => {
      // #region covers 584adc09feb4d383

      // We can grab the full handler type from the Schema:
      type myQueryHandler = Schema['myQuery']['functionHandler'];

      // For demonstration, we'll also grab the args and return types:
      type myQueryArgs = Schema['myQuery']['args'];
      type myQueryResult = Schema['myQuery']['returnType'];

      // We can then define a function handler, which will normally will be defined
      // in your `amplify/` folder, and we can add our `myQueryHandler` type:
      const handler: myQueryHandler = async (event, context) => {
        // The `event` arg is now fully typed. And `event.arguments` will match
        // the `args` type from the custom operation.
        const args: myQueryArgs = event.arguments;

        // We can build our result using the result type:
        const result: myQueryResult = {
          // ctrl + space in VS Code auto-suggests `output: ...`
          output: 'some output',
        };

        // And we can return `result`, which matches the output demanded by
        // the function signature.
        return result;
      };

      // #endregion
    });
  });
});
