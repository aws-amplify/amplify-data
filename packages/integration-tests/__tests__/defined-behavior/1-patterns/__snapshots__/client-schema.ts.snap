// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ClientSchema overview sample schema 1`] = `
"enum CustomerEngagementStage {
  PROSPECT
  INTERESTED
  PURCHASED
}

type Customer @model @auth(rules: [{allow: public, provider: apiKey}])
{
  customerId: ID! @primaryKey
  name: String
  location: CustomerLocation
  engagementStage: CustomerEngagementStage
  collectionId: ID
  collection: Collection @belongsTo(references: ["collectionId"])
}

type Collection @model @auth(rules: [{allow: public, provider: apiKey}])
{
  customers: [Customer] @hasMany(references: ["collectionId"])
  tags: [String]
  representativeId: ID! @index(queryField: "listCollectionByRepresentativeId")
}

type CustomerLocation 
{
  lat: Float!
  long: Float!
}"
`;
